kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java

```diff
@@ -36,23 +40,34 @@ public class SnapshotImpl implements Snapshot {
     private final long version;
     private final LogReplay logReplay;
     private final Lazy<Tuple2<Protocol, Metadata>> protocolAndMetadata;
+    private final Optional<SnapshotHint> snapshotHint;
 
     public SnapshotImpl(
-        Path logPath,
-        Path dataPath,
-        long version,
-        LogSegment logSegment,
-        TableClient tableClient,
-        long timestamp) {
+            Path logPath,
+            Path dataPath,
+            long version,
+            LogSegment logSegment,
+            TableClient tableClient,
+            long timestamp,
+            SnapshotManager snapshotManager,
+            Optional<SnapshotHint> snapshotHint) {
         this.dataPath = dataPath;
         this.version = version;
-
         this.logReplay = new LogReplay(
             logPath,
             dataPath,
             tableClient,
-            logSegment);
-        this.protocolAndMetadata = new Lazy<>(logReplay::loadProtocolAndMetadata);
+            logSegment,
+            snapshotHint);
+        this.protocolAndMetadata = new Lazy<>(() -> {
+            // Construct the SnapshotHint lazily. i.e. only create it when some caller/consumer of
+            // this SnapshotImpl instance decides to load the protocol and metadata.
+            final Tuple2<Protocol, Metadata> pAndM = logReplay.loadProtocolAndMetadata();
+            final SnapshotHint hint = new SnapshotHint(version, pAndM._1, pAndM._2);
+            snapshotManager.registerHint(hint);
+            return pAndM;
+        });
+        this.snapshotHint = snapshotHint;
```

Comment:
Why does this need to be a member? The value gets used by the lazy computation, but this field has no use sites?